data/scenarios/Challenges/hanoi-invariant.sw

```diff
@@ -0,0 +1,51 @@
+def repeat = \c. c; repeat c end;
+def isUnlocked = \e. e == "one" || e == "two" || e == "three" end;
+def unlock = \e.
+  if (e == "blocked one")   {"one"}   {
+  if (e == "blocked two")   {"two"}   {
+  if (e == "blocked three") {"three"} {
+  fail $ "Can not unlock: " ++ e
+  }}}
+end;
+
+repeat (
+me <- scan down;
+case me (\_. return ()) (\e.
+// if
+//  0. I stand on unlocked X
+//  1. place north of me is NOT empty
+// then
+//  - lock X
```

Comment:
While playing around I found another condition:

If:
- I stand on an unlocked disk
- the count of all placed is not 3

Then:
- I should lock the disk to preserve the invariant

Reply:
How is it possible to get into that state though?

Reply:
If you have for example:
```
| | | |
| | | |
|1|2|3|
+-+-+-+
```
Then at that moment all of them will be unlocked. Which makes sense (except for the 3).

Once you grab any of them, the others should immediately be locked, so that you can not grab more and swap them.

Reply:
Ah, I see!  You're right, I can confirm that at the moment it is possible to cheat: put them as in the picture you drew above, then pick them all up, move to the last column, and place them going up the column one after another.